velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp

```diff
@@ -0,0 +1,169 @@
+      : client_(std::make_unique<DataLakeFileClient>(client)) {}
+
+  void create() override {
+    client_->Create();
+  }
+
+  PathProperties getProperties() override {
+    return client_->GetProperties().Value;
+  }
+
+  void append(const uint8_t* buffer, size_t size, uint64_t offset) override {
+    auto bodyStream = Azure::Core::IO::MemoryBodyStream(buffer, size);
+    client_->Append(bodyStream, offset);
+  }
+
+  void flush(uint64_t position) override {
+    client_->Flush(position);
+  }
+
+  void close() override {
+    // do nothing.
+  }
+
+ private:
+  std::unique_ptr<DataLakeFileClient> client_;
+};
+
+class AbfsWriteFile::Impl {
+ public:
+  explicit Impl(const std::string& path, const std::string& connectStr)
+      : path_(path), connectStr_(connectStr) {
+    // Make it a no-op if invoked twice.
+    if (position_ != -1) {
+      return;
+    }
+    position_ = 0;
+  }
+
+  void initialize() {
+    if (!blobStorageFileClient_) {
+      auto abfsAccount = AbfsAccount(path_);
+      auto fileClient = DataLakeFileClient::CreateFromConnectionString(
+          connectStr_, abfsAccount.fileSystem(), abfsAccount.filePath());
+      blobStorageFileClient_ = std::make_unique<BlobStorageFileClient>(
+          BlobStorageFileClient(fileClient));
+    }
+
+    VELOX_CHECK(!exist(), "File already exists");
+    blobStorageFileClient_->create();
+  }
+
+  /// mainly for test purpose.
+  void setFileClient(
+      std::shared_ptr<IBlobStorageFileClient> blobStorageManager) {
+    blobStorageFileClient_ = std::move(blobStorageManager);
+  }
+
+  bool exist() {
+    try {
+      blobStorageFileClient_->getProperties();
+      return true;
+    } catch (Azure::Storage::StorageException& e) {
+      if (e.StatusCode == Azure::Core::Http::HttpStatusCode::NotFound) {
+        return false;
+      } else {
+        throwStorageExceptionWithOperationDetails("GetProperties", path_, e);
+      }
+    }
+  }
+
+  void close() {
+    if (!closed_) {
+      flush();
+      blobStorageFileClient_->close();
+      closed_ = true;
+    }
+  }
+
+  void flush() {
+    if (!closed_) {
+      blobStorageFileClient_->flush(position_);
+    }
+  }
+
+  void append(std::string_view data) {
+    VELOX_CHECK(!closed_, "File is not open");
+    if (data.size() == 0) {
+      return;
+    }
+    append(data.data(), data.size());
+  }
+
+  uint64_t size() const {
+    auto properties = blobStorageFileClient_->getProperties();
+    return properties.FileSize;
+  }
+
+  void append(const char* buffer, size_t size) {
+    auto offset = position_;
+    position_ += size;
+    blobStorageFileClient_->append(
+        reinterpret_cast<const uint8_t*>(buffer), size, offset);
+  }
+
+ private:
+  const std::string path_;
+  const std::string connectStr_;
+  std::string fileSystem_;
+  std::string fileName_;
+  std::shared_ptr<IBlobStorageFileClient> blobStorageFileClient_;
+
+  uint64_t position_ = -1;
+  std::atomic<bool> closed_{false};
```

Comment:
I don't think atomic is needed here since the assumption is that `AbfsWriteFile` is not thread-safe.